x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportUpdateTrainedModelAliasAction.java

```diff
@@ -0,0 +1,237 @@
+                }
+                if (request.getOldModelId() != null) {
+                    if (oldModel == null) {
+                        listener.onFailure(
+                            ExceptionsHelper.badRequestException("cannot find model matching old_model_id [{}]", request.getOldModelId())
+                        );
+                        return;
+                    }
+                    if (isLicensed.test(oldModel) == false) {
+                        listener.onFailure(LicenseUtils.newComplianceException(XPackField.MACHINE_LEARNING));
+                        return;
+                    }
+
+                    // validate inference configs are the same type. Moving an alias from regression -> classification seems dangerous
+                    if (newModel.getInferenceConfig() != null && oldModel.getInferenceConfig() != null) {
+                        if (newModel.getInferenceConfig().getName().equals(oldModel.getInferenceConfig().getName()) == false) {
+                            listener.onFailure(
+                                ExceptionsHelper.badRequestException(
+                                    "new_model_id [{}] has inference config type [{}] but old_model_id [{}] has type [{}]",
+                                    newModel.getModelId(),
+                                    newModel.getInferenceConfig().getName(),
+                                    oldModel.getModelId(),
+                                    oldModel.getInferenceConfig().getName()
+                                )
+                            );
+                            return;
+                        }
+                    }
+
+                    Set<String> oldInputFields = new HashSet<>(oldModel.getInput().getFieldNames());
+                    Set<String> newInputFields = new HashSet<>(newModel.getInput().getFieldNames());
+                    // TODO should we fail in this case???
+                    if (Sets.difference(oldInputFields, newInputFields).size() > (oldInputFields.size() / 2)
+                    || Sets.intersection(newInputFields, oldInputFields).size() < (oldInputFields.size() / 2)) {
+                        String warning =  Messages.getMessage(
+                            TRAINED_MODEL_INPUTS_DIFFER_SIGNIFICANTLY,
+                            UpdateTrainedModelAliasAction.Request.OLD_MODEL_ID.getPreferredName(),
+                            request.getOldModelId(),
+                            UpdateTrainedModelAliasAction.Request.NEW_MODEL_ID.getPreferredName(),
+                            request.getNewModelId());
+                        auditor.warning(request.getOldModelId(), warning);
+                        logger.warn("[{}] {}", request.getOldModelId(), warning);
+                        HeaderWarning.addWarning(warning);
+                    }
+                }
+                clusterService.submitStateUpdateTask("update-model-alias", new AckedClusterStateUpdateTask(request, listener) {
```

Comment:
I thought we "got this for free" when we set the minimum supported node version in the custom metadata.

Reply:
@davidkyle this is a master node action. I think that + the minimum supported node version in the metadata is enough.

Digging into `getMinimalSupportedVersion()` it indeed puts a barrier on writing out the custom metadata to the stream. It will not be written to older nodes that don't understand it.

Reply:
👍 we are good once the master node has been upgraded